src/cluster_undersample/cluster_undersample.jl

```diff
@@ -0,0 +1,207 @@
+    seed!(rng)               # kmeans offers no better way :(
+    result = kmeans(X, n; maxiter)
+    # the n cluster centers are the undersampled points
+    X_new = result.centers
+    # unless mode is "nearest" where we seek the nearest neighbor of each center
+    if mode == "nearest"
+        tree = BallTree(X)
+        keep_inds, _ = knn(tree, X_new, 1, true)
+        keep_inds = vcat(keep_inds...)      # flatten
+        X_new = X[:, keep_inds]
+    end
+    return X_new
+end
+
+"""
+    cluster_undersample(
+        X, y; 
+        mode= "nearest", ratios = 1.0, maxiter = 100,
+        rng=default_rng(), try_preserve_type=true
+    )
+
+
+# Description
+
+Undersample a dataset using clustering undersampling as presented in [1] using K-means as the clustering algorithm.
+
+# Positional Arguments
+
+$(COMMON_DOCS["INPUTS"])
+
+# Keyword Arguments
+
+- `mode::AbstractString="nearest`: If `center` then the undersampled data will consist of the centriods of 
+    each cluster found. Meanwhile, if `nearest` then it will consist of the nearest neighbor of each centroid.
+
+$(COMMON_DOCS["RATIOS-UNDERSAMPLE"])
+
+- `maxiter::Integer=100`: Maximum number of iterations to run K-means
+
+- `rng::Integer=42`: Random number generator seed. Must be an integer.
+
+$(COMMON_DOCS["TRY_PERSERVE_TYPE"])
+
+# Returns
+
+$(COMMON_DOCS["OUTPUTS-UNDER"])
+
+
+# Example
+```julia
+using Imbalance
+
+# set probability of each class
+probs = [0.5, 0.2, 0.3]                         
+num_rows, num_continuous_feats = 100, 5
+# generate a table and categorical vector accordingly
+X, y = generate_imbalanced_data(num_rows, num_continuous_feats; 
+                                probs, rng=42)                       
+julia> checkbalance(y; ref="minority")
+ 1: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 19 (100.0%) 
+ 2: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 33 (173.7%) 
+ 0: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 48 (252.6%) 
+
+# apply cluster_undersampling
+X_under, y_under = cluster_undersample(X, y; mode="nearest", 
+                                       ratios=Dict(0=>1.0, 1=> 1.0, 2=>1.0), rng=42)
+julia> checkbalance(y_under)
+0: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 19 (100.0%) 
+2: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 19 (100.0%) 
+1: ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 19 (100.0%) 
+```
+
+# MLJ Model Interface
+
+Simply pass the keyword arguments while initiating the `ClusterUndersampler` model and pass the 
+    positional arguments to the `transform` method. 
+
+```julia
+using MLJ
+ClusterUndersampler = @load ClusterUndersampler pkg=Imbalance
+
+# Wrap the model in a machine
+undersampler = ClusterUndersampler(mode="nearest", ratios=Dict(0=>1.0, 1=> 1.0, 2=>1.0), rng=42)
+mach = machine(undersampler)
+
+# Provide the data to transform (there is nothing to fit)
+X_under, y_under = transform(mach, X, y)
+```
+The `MLJ` interface is only supported for table inputs. Read more about the interface [here]().
```

Comment:
> The `MLJ` interface is only supported for table inputs.

Is that true? The `input_scitype` declaration in [src/cluster_undersample/interface_mlj.jl](https://github.com/JuliaAI/Imbalance.jl/pull/56/files#diff-0249c3c1263a3d8055accfa726f196af57390931209d125aa09f556e5de5929a) suggests matrices are also supported.

Reply:
No it is not. Let me go back and fix that. However, it may not be a lot of effort per interface to also make a dispatch for matrix input so we can consider that later if you like.

Reply:
Okay I have just finished that. Now MLJ interface with matrices is also supported for all 8 of the 9 methods (had good reason to exclude SMOTE-NC and made it clear in docs).

Reply:
👍🏾